lighthouse-core/gather/gatherers/scripts.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * @license Copyright 2022 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const FRGatherer = require('../../fraggle-rock/gather/base-gatherer.js');
+
+/**
+ * @template T, U
+ * @param {Array<T>} values
+ * @param {(value: T) => Promise<U>} promiseMapper
+ * @param {boolean} runInSeries
+ * @return {Promise<Array<U>>}
+ */
+async function runInSeriesOrParallel(values, promiseMapper, runInSeries) {
+  if (runInSeries) {
+    const results = [];
+    for (const value of values) {
+      const result = await promiseMapper(value);
+      results.push(result);
+    }
+    return results;
+  } else {
+    const promises = values.map(promiseMapper);
+    return await Promise.all(promises);
+  }
+}
+
+/**
+ * @fileoverview Gets JavaScript file contents.
+ */
+class Scripts extends FRGatherer {
+  /** @type {LH.Gatherer.GathererMeta} */
+  meta = {
+    supportedModes: ['timespan', 'navigation'],
+  };
+
+  /** @type {LH.Crdp.Debugger.ScriptParsedEvent[]} */
+  _scriptParsedEvents = [];
+
+  /** @type {Array<string | undefined>} */
+  _scriptContents = [];
+
+  /** @type {string|null|undefined} */
+  _mainSessionId = null;
+
+  constructor() {
+    super();
+    this.onProtocolMessage = this.onProtocolMessage.bind(this);
+  }
+
+  /**
+   * @param {LH.Protocol.RawEventMessage} event
+   */
+  onProtocolMessage(event) {
+    // Go read the comments in network-recorder.js _findRealRequestAndSetSession.
+    let sessionId = event.sessionId;
+    if (this._mainSessionId === null) {
+      this._mainSessionId = sessionId;
+    }
+    if (this._mainSessionId === sessionId) {
+      sessionId = undefined;
+    }
+
+    // We want to ignore scripts from OOPIFs. In reality, this does more than block just OOPIFs,
+    // it also blocks scripts from the same origin but that happen to run in a different process,
+    // like a worker.
+    if (event.method === 'Debugger.scriptParsed' && !sessionId) {
+      // Events without an embedderName (read: a url) are for JS that we ran over the protocol.
+      if (event.params.embedderName) this._scriptParsedEvents.push(event.params);
```

Comment:
Nit: could just move this condition to the outer if statement

Reply:
It's layered this way b/c comments.